### ISSUE TITLE
[BUG] loss can not decrease when using overlap-param-gather

### ISSUE THREAD

**Issue by jindajia** [NVIDIA/Megatron-LM#663]
**Describe the bug**
Hello,

I've encountered an issue where using `overlap-gather-parameters` and `overlap-reduce-gradients` simultaneously seems to prevent the loss from decreasing as expected. It appears that in the `optimizer.step()` function, the `found_inf_flag` is consistently true, leading to the learning rate remaining at zero.

To investigate this, I experimented with three different settings:
1. Without overlapping both gradient reduction and parameter gathering, the loss decreases normally.
2. With overlapping gradient reduction but without overlapping parameter gathering, the loss still decreases normally.
3. However, when both gradient reduction and parameter gathering are overlapped, the loss does not decrease.

It seems there might be an issue with the overlapping parameter gathering. Any insights or suggestions on this matter would be greatly appreciated.

**To Reproduce**
Steps to reproduce the behavior. The easier it is to reproduce the faster it will get maintainer attention.
Training Arguments:
```
MODEL_ARGS="
    --num-layers 12 \
    --hidden-size 768 \
    --num-attention-heads 12 \
    --seq-length 1024 \
    --max-position-embeddings 1024 \
"

OPTIMIZER_ARGS="
    --lr 0.0006 \
    --lr-decay-iters 70000 \
    --lr-decay-style cosine \
    --min-lr 0.00006 \
    --adam-beta1 0.9 \
    --adam-beta2 0.95 \
    --adam-eps 1e-08 \
    --weight-decay .1 \
    --lr-warmup-fraction 0.01 \
    --clip-grad 1.0 \
    --fp16 \
    --use-distributed-optimizer \
"

TRAING_ARGS="
    --tensor-model-parallel-size 1 \
    --pipeline-model-parallel-size 1 \
    --micro-batch-size 2 \
    --global-batch-size 256 \
    --train-iters 80000 \
"

OUTPUT_ARGS="
    --log-interval 10 \
    --timing-log-level 2 \
    --save-interval 1000 \
    --eval-interval 100 \
    --eval-iters 10 \
    --log-timers-to-tensorboard \
    --log-validation-ppl-to-tensorboard \
    --log-throughput \
    --tensorboard-dir ${TENSORBOARD_DIR} \
    --tensorboard-log-interval 1 \
    --wandb-project zeropp-quantization\
    --wandb-save-dir ${WANDB_DIR} \
    --wandb-exp-name qt-gradients \
"

PERFORMANCE_ARGS="
    --no-async-tensor-model-parallel-allreduce \
    --recompute-activations \
    --recompute-granularity selective \
    --overlap-param-gather \
    --overlap-grad-reduce \
    --no-delay-grad-reduce \
"
```

**Environment (please complete the following information):**
 - Megatron-LM commit ID: a22f6723
 - PyTorch version 2.0.1
 - CUDA 11.8
 - NCCL version 2.14.3

**Additional context**
output log
```
(min, max) time across ranks (ms):
    load-checkpoint ................................: (0.28, 26.66)
/ocean/projects/asc200010p/jjia1/TOOLS/mambaforge/envs/megatron-cuda_118/lib/python3.10/site-packages/torch/utils/data/dataloader.py:560: UserWarning: This DataLoader will create 2 worker processes in total. Our suggested max number of worker in current system is 1, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
(min, max) time across ranks (ms):
    model-and-optimizer-setup ......................: (416.54, 452.18)
    train/valid/test-data-iterators-setup ..........: (4135.60, 4811.10)
 iteration       10/   80000 | consumed samples:         2560 | elapsed time per iteration (ms): 4985.7 | throughput per GPU (TFLOP/s/GPU): 11.2 | learning rate: 0.000E+00 | global batch size:   256 | loss scale: 8388608.0 | number of skipped iterations:  10 | number of nan iterations:   0 |
(min, max) time across ranks (ms):
    forward-backward ...............................: (4687.59, 4725.48)
    forward-compute ................................: (2526.55, 2652.14)
    backward-compute ...............................: (2044.25, 2154.34)
    batch-generator ................................: (124.27, 148.14)
    layernorm-grads-all-reduce .....................: (0.03, 0.04)
    embedding-grads-all-reduce .....................: (0.05, 0.05)
    all-grads-sync .................................: (0.05, 0.06)
    optimizer-copy-to-main-grad ....................: (0.82, 2.41)
    optimizer-unscale-and-check-inf ................: (16.52, 26.60)
    optimizer ......................................: (56.70, 65.18)
 iteration       20/   80000 | consumed samples:         5120 | elapsed time per iteration (ms): 3647.3 | throughput per GPU (TFLOP/s/GPU): 15.4 | learning rate: 0.000E+00 | global batch size:   256 | loss scale: 8192.0 | number of skipped iterations:  10 | number of nan iterations:   0 |
(min, max) time across ranks (ms):
    forward-backward ...............................: (3394.48, 3425.06)
    forward-compute ................................: (1937.33, 2043.74)
    backward-compute ...............................: (1344.54, 1476.60)
    batch-generator ................................: (30.67, 35.40)
    layernorm-grads-all-reduce .....................: (0.03, 0.04)
    embedding-grads-all-reduce .....................: (0.04, 0.05)
    all-grads-sync .................................: (0.05, 0.05)
    optimizer-copy-to-main-grad ....................: (0.76, 1.83)
    optimizer-unscale-and-check-inf ................: (15.69, 23.80)
    optimizer ......................................: (48.94, 60.48)
 iteration       30/   80000 | consumed samples:         7680 | elapsed time per iteration (ms): 3625.5 | throughput per GPU (TFLOP/s/GPU): 15.4 | learning rate: 0.000E+00 | global batch size:   256 | loss scale: 8.0 | number of skipped iterations:  10 | number of nan iterations:   0 |
(min, max) time across ranks (ms):
    forward-backward ...............................: (3377.82, 3393.75)
    forward-compute ................................: (1951.96, 1993.98)
    backward-compute ...............................: (1376.64, 1430.05)
    batch-generator ................................: (25.03, 29.31)
    layernorm-grads-all-reduce .....................: (0.03, 0.03)
    embedding-grads-all-reduce .....................: (0.04, 0.05)
    all-grads-sync .................................: (0.05, 0.05)
    optimizer-copy-to-main-grad ....................: (0.76, 1.54)
    optimizer-unscale-and-check-inf ................: (13.68, 20.10)
    optimizer ......................................: (50.34, 56.82)
 iteration       40/   80000 | consumed samples:        10240 | elapsed time per iteration (ms): 3621.3 | throughput per GPU (TFLOP/s/GPU): 15.5 | learning rate: 0.000E+00 | global batch size:   256 | loss scale: 1.0 | number of skipped iterations:  10 | number of nan iterations:   0 |
(min, max) time across ranks (ms):
    forward-backward ...............................: (3383.90, 3396.33)
    forward-compute ................................: (1940.77, 1980.24)
    backward-compute ...............................: (1407.62, 1447.45)
    batch-generator ................................: (24.00, 32.65)
    layernorm-grads-all-reduce .....................: (0.03, 0.03)
    embedding-grads-all-reduce .....................: (0.04, 0.05)
    all-grads-sync .................................: (0.04, 0.05)
    optimizer-copy-to-main-grad ....................: (0.83, 1.47)
    optimizer-unscale-and-check-inf ................: (15.45, 27.31)
    optimizer ......................................: (51.82, 55.86)
 iteration       50/   80000 | consumed samples:        12800 | elapsed time per iteration (ms): 3643.5 | throughput per GPU (TFLOP/s/GPU): 15.4 | learning rate: 0.000E+00 | global batch size:   256 | loss scale: 1.0 | number of skipped iterations:  10 | number of nan iterations:   0 |
(min, max) time across ranks (ms):
    forward-backward ...............................: (3397.56, 3422.50)
    forward-compute ................................: (1938.00, 2043.37)
    backward-compute ...............................: (1342.57, 1452.97)
    batch-generator ................................: (24.89, 35.75)
    layernorm-grads-all-reduce .....................: (0.03, 0.03)
    embedding-grads-all-reduce .....................: (0.04, 0.05)
    all-grads-sync .................................: (0.05, 0.06)
    optimizer-copy-to-main-grad ....................: (0.78, 2.44)
    optimizer-unscale-and-check-inf ................: (11.96, 22.84)
    optimizer ......................................: (45.09, 56.44)
```

**Comment by deepakn94** [comment]
Will look into this.

**Comment by jindajia** [comment]
Thank you! Have you reproduced this issue?

**Comment by deepakn94** [comment]
I just reproduced it, thanks. Trying to figure out what's wrong, and then will implement a fix.

**Comment by deepakn94** [comment]
I have a fix, it just needs to get through internal peer review and then it will be pushed here.

**Comment by jindajia** [comment]
Awesome, thank you so much. In the past few days, I also found an informal way to bypass this bug, which is to manually set the `--initial-loss-scale`. This can also temporarily circumvent the issue.

**Comment by deepakn94** [comment]
@jindajia: The fix has been pushed (https://github.com/NVIDIA/Megatron-LM/commit/75120dbf069703d21e34f0cb031fb4a3c1ac5315). Please let me know if you are still running into issues.

**Comment by jindajia** [comment]
Thank you for your repair!! I will try it and let you know if there are anything issues.

**Comment by uygnef** [comment]
@deepakn94 Hi, I encountered an issue. I set --overlap-grad-reduce and --overlap-param-gather.

I use both train and eval. After eval finishes, the first step of the next train has a gradient of 0, but the subsequent train steps have normal gradients. 
Do you have any ideas? If I set force_sync=True here, the gradients become normal.
